Fixing inconsistency between generated entropy value type and the expected HDWallet.entropy value type

daehan-koreapool opened the pull request:
`mnemonic` Python package generates an entropy value in `bytearray` type.
However, `HDWallet.entropy` class is expecting a string value and it is used as a string throughout the entire bip32.py module codebase.

This merge request explicitly converts generated entropy value as a string value to be consistent with the original implementer's intention.

daehan-koreapool commented:
```
-> print(entropy)
(Pdb) entropy
bytearray(b'\xdf\x9e\xd2^\xd1F\xbfC3j]|\xf79Y\x94')
(Pdb) unhexlify(entropy)
*** binascii.Error: Non-hexadecimal digit found
```
Apparently `entropy = Mnemonic(language="english").to_entropy(words=mnemonic)` code doesn't seem to be hex decodable binary data.

Storing `HDWallet.entropy` in bytes might be the only solution here.

daehan-koreapool commented on pycardano/crypto/bip32.py:
```diff
@@ -152,28 +155,18 @@ def from_mnemonic(cls, mnemonic: str, passphrase: str = "") -> HDWallet:
             raise ValueError("Invalid mnemonic words.")
 
         mnemonic = unicodedata.normalize("NFKD", mnemonic)
-        passphrase = str(passphrase) if passphrase else ""
```
No need to convert it to a string when the method parameter already explicitly requires an optional string type.

cffls commented:
> Since crypto/bip32.py is being implemented and maintained by PyCardano project contributers, I think this module should be included in the coverage report in my opinion.
> 
> What do you think @cffls ?

Thank you for raising this PR! Yes, I think it makes sense to include this file to coverage. Please feel free to change `.coveragerc`.

daehan-koreapool commented:
# Major Updates
* Ensuring Mnemonic package generated entropy value is stored as a string representation
* Pull out seed generating method for reusability
* Setting `from_entropy()` passphrase parameter as a string value by default
* Add more testcases ported from Emurgo serialization library to ensure both Mainnet and Testnet addresses are covered
* Add tests around entropy-based class methods to ensure refactored codes are functional

# Minor Updates
* Replace print statements with logging statements
* Update Makefile, gitignore, and coveragerc

cffls reviewed (approved):
LGTM, thanks for contributing to this project! @henryyuanheng-wang Henry, could you please take a look at this?

daehan-koreapool commented:
Hi @henryyuanheng-wang ,

I have a question regarding `test_payment_address_12_reward2()` test case.
For this test, I see that you are explicitly specifying `private=False` for the last two derivation path indexes.
From my experiment, setting this to `private=True` results in the same public key generation.
Would you be able to clarify on why this results in the same public key?

I've also just pushed another commit which can be rolled back if you wish to.
This commit is to enhance HDWallet creation UX to support simpler chaining of HDWallet derivation steps which is similar to other HDWallet library API design.

henryyuanheng-wang commented:
> Hi @henryyuanheng-wang ,
> 
> I have a question regarding `test_payment_address_12_reward2()` test case. For this test, I see that you are explicitly specifying `private=False` for the last two derivation path indexes. From my experiment, setting this to `private=True` results in the same public key generation. Would you be able to clarify on why this results in the same public key?
> 
> I've also just pushed another commit which can be rolled back if you wish to. This commit is to enhance HDWallet creation UX to support simpler chaining of HDWallet derivation steps which is similar to other HDWallet library API design.

Hi @daehan-koreapool , thanks for catching the inconsistencies and for making some great improvements! Everything looks pretty good. I like how you simplified the derivation step so a new HDWallet instance is created without needing to pass a parent HDWallet (was a bit redundant in hindsight). One question regarding the function name change from `derive_from_index` to `derive` - can you point me to the other popular implementation you were referring to? I was referencing some of the naming conventions from `python-hdwallet` [repo](https://github.com/meherett/python-hdwallet) but would love to see some other implementations. 

To answer your question regarding public keys derived either `private` or `public` is the same - I think this is expected behavior and beauty of HDWallet is that with public key from account you can derive the same public address as if you had the private key (you can refer to the diagram [here](https://input-output-hk.github.io/cardano-wallet/concepts/address-derivation) under HD Sequential wallets section and a HDWallet reading I found really useful [here](https://learnmeabitcoin.com/technical/hd-wallets)).  Let me know if it makes sense!

daehan-koreapool commented:
Hi @henryyuanheng-wang,

> Hi @daehan-koreapool , thanks for catching the inconsistencies and for making some great improvements! Everything looks pretty good. I like how you simplified the derivation step so a new HDWallet instance is created without needing to pass a parent HDWallet (was a bit redundant in hindsight). One question regarding the function name change from `derive_from_index` to `derive` - can you point me to the other popular implementation you were referring to? I was referencing some of the naming conventions from `python-hdwallet` [repo](https://github.com/meherett/python-hdwallet) but would love to see some other implementations.

It's pretty cool to know that derived public keys are identical for both private and public derivation methods.
Thanks for the clarification.

> To answer your question regarding public keys derived either `private` or `public` is the same - I think this is expected behavior and beauty of HDWallet is that with public key from account you can derive the same public address as if you had the private key (you can refer to the diagram [here](https://input-output-hk.github.io/cardano-wallet/concepts/address-derivation) under HD Sequential wallets section and a HDWallet reading I found really useful [here](https://learnmeabitcoin.com/technical/hd-wallets)). Let me know if it makes sense!

Regarding the `derive()` naming update, I referred to [`cardano-serialization-lib`](https://github.com/Emurgo/cardano-serialization-lib/blob/master/rust/src/address.rs#L1019) rust implementation.
This API design seem very simple and intuitive to use, so here's where I got the idea of returning a new HDWallet each time a derive() method is called for easier chaining operations.

[Lucid js library](https://doc.deno.land/https://deno.land/x/lucid@0.6.7/mod.ts/~/Core.Bip32PrivateKey
) also follows similar `derive(index)` pattern but this could be because Lucid uses `cardano-serialization-lib` wasm module under the hood.

daehan-koreapool commented:
I think the coverage test is failing since we are now including bip32.py module in the coverage report.
I'll add more testcases try to increase the test coverage.

daehan-koreapool commented:
Added a bit more refactoring commits as well.
* Pulled out frequently used list as a constant set. This could be set as a tuple if we want immutability, but since we use "contains" check more often, thought a set data structure is more suitable
* Cleaned up nested statements
